lib/libzpool/vdev_disk_aio.c

```diff
@@ -404,6 +408,55 @@ kick_submitter(vdev_disk_aio_t *vda)
 	assert(rc == sizeof (data));
 }
 
+#define	SYNCHRONIZE_CACHE_CMD	0x35
+#define	SYNCHRONIZE_CACHE_CMDLEN	10
+#define	FLUSH_TIMEOUT	1000
+
+/*
+ * This flush write-cache function works only for true SCSI disks (sd driver):
+ *
+ *  *) NVMe devices don't support the ioctl,
+ *  *) ATA/SATA disks haven't been tested.
+ *
+ * NOTE: This is called synchronously in zio pipeline. Attempt to execute
+ * flush asynchronously on behalf of taskq thread resulted in -10%
+ * performance regression for sync workloads.
+ */
+static void
+vdev_disk_aio_flush(zio_t *zio)
+{
+	vdev_t *vd = zio->io_vd;
+	vdev_disk_aio_t *vda = vd->vdev_tsd;
+
+	struct sg_io_hdr io_hdr;
+	unsigned char scCmdBlk[SYNCHRONIZE_CACHE_CMDLEN] =
+	    {SYNCHRONIZE_CACHE_CMD, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+	unsigned char sense_b[32];
+
+	memset(&io_hdr, 0, sizeof (io_hdr));
+
+	io_hdr.interface_id = 'S';
+	io_hdr.cmd_len = SYNCHRONIZE_CACHE_CMDLEN;
+	io_hdr.cmdp = scCmdBlk;
+	io_hdr.sbp = sense_b;
+	io_hdr.mx_sb_len = sizeof (sense_b);
+	io_hdr.dxfer_direction = SG_DXFER_NONE;
+	io_hdr.timeout = FLUSH_TIMEOUT;
+
+	if (ioctl(vda->vda_fd, SG_IO, &io_hdr) < 0 ||
+	    io_hdr.status != 0) {
+		if (errno == EINVAL || errno == ENOTTY) {
+			fprintf(stderr, "Disk %s does not support synchronize "
+			    "cache SCSI command\n", vd->vdev_path);
+			vda->vda_noflush = B_TRUE;
```

Comment:
Can dump the sense data in case of failure? Depending on the failed mode -- we might as well disable flushing altogether for this device for future IOs?

Reply:
I try to distinguish between two cases:

1) cache flush not possible
       * this is either EINVAL if SG_IO is not supported at all (i.e. lofi device)
       * or ENOTTY if SG_IO is understood nevertheless cache synchronize command not implemented
2) transient error (all other cases)

For case number 1 we never attempt to flush again in the future (stop losing time). For case number 2 we return error for that particular zio, but cache flushing remains enabled.

Regarding printing sense data, there is a bug in current fix as I need to distinguish between failed ioctl and failed SCSI command. For failed SCSI command we could print the sense data to stderr. I would prefer to print it raw (in hex), the routine for decoding sense data would be pretty long (example: https://fossies.org/dox/sg3_utils-1.42/sg__lib_8c_source.html#l01412) . So everytime when ioctl succeeds but SCSI cmd fails we will print an error to stderr. I just hope it won't happen too frequently :-) 

Let me know if you agree with the proposal above. thanks